ee/clickhouse/models/action.py

```diff
@@ -19,55 +18,65 @@ def format_action_filter(action: Action, prepend: str = "", index=0) -> Tuple[st
 
     or_queries = []
     for index, step in enumerate(steps):
+        conditions: List[str] = []
         # filter element
         if step.event == AUTOCAPTURE_EVENT:
-            query, element_params, index = filter_element(step, "{}{}".format(index, prepend), index)
+            el_conditions, element_params = filter_element(step, "{}{}".format(index, prepend))
             params = {**params, **element_params}
-        # filter event
-        else:
-            query, event_params, index = filter_event(step, "{}{}".format(index, prepend), index)
-            params = {**params, **event_params}
+            conditions += el_conditions
+
+        # filter event conditions (ie URL)
+        event_conditions, event_params = filter_event(step, "{}{}".format(index, prepend), index)
+        params = {**params, **event_params}
+        conditions += event_conditions
 
         if step.properties:
             from ee.clickhouse.models.property import parse_prop_clauses
 
             prop_query, prop_params = parse_prop_clauses(
                 "uuid", Filter(data={"properties": step.properties}).properties, action.team
             )
-            query += prop_query
+            conditions.append(prop_query.replace("AND", "", 1))
             params = {**params, **prop_params}
 
-        or_queries.append(query)
-    or_separator = "OR uuid IN"
-    formatted_query = or_separator.join(or_queries)
-
+        if len(conditions) > 0:
+            or_queries.append(" AND ".join(conditions))
+    or_separator = (
+        ") OR (" if not use_loop else ") OR uuid IN (SELECT uuid FROM events WHERE team_id = %(team_id)s AND "
+    )
+    if use_loop:
```

Comment:
NIT: for clarity can just put the or_separator under the respective if condition. The ternary above is the same as this if statement

Reply:
Done!